examples/platforms/emsk/radio.c

```diff
@@ -0,0 +1,660 @@
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the OpenThread platform abstraction for radio communication.
+ *
+ */
+
+#include "openthread/types.h"
+
+#include <common/code_utils.hpp>
+#include "openthread/platform/radio.h"
+#include "platform-emsk.h"
+
+#include "device/device_hal/inc/dev_gpio.h"
+
+#include <stdio.h>
+#define DBG(fmt, ...)   printf(fmt, ##__VA_ARGS__)
+
+enum
+{
+    IEEE802154_MIN_LENGTH         = 5,
+    IEEE802154_MAX_LENGTH         = 127,
+    IEEE802154_ACK_LENGTH         = 5,
+
+    IEEE802154_BROADCAST          = 0xffff,
+
+    IEEE802154_FRAME_TYPE_ACK     = 2 << 0,
+    IEEE802154_FRAME_TYPE_MACCMD  = 3 << 0,
+    IEEE802154_FRAME_TYPE_MASK    = 7 << 0,
+
+    IEEE802154_SECURITY_ENABLED   = 1 << 3,
+    IEEE802154_FRAME_PENDING      = 1 << 4,
+    IEEE802154_ACK_REQUEST        = 1 << 5,
+    IEEE802154_PANID_COMPRESSION  = 1 << 6,
+
+    IEEE802154_DST_ADDR_NONE      = 0 << 2,
+    IEEE802154_DST_ADDR_SHORT     = 2 << 2,
+    IEEE802154_DST_ADDR_EXT       = 3 << 2,
+    IEEE802154_DST_ADDR_MASK      = 3 << 2,
+
+    IEEE802154_SRC_ADDR_NONE      = 0 << 6,
+    IEEE802154_SRC_ADDR_SHORT     = 2 << 6,
+    IEEE802154_SRC_ADDR_EXT       = 3 << 6,
+    IEEE802154_SRC_ADDR_MASK      = 3 << 6,
+
+    IEEE802154_DSN_OFFSET         = 2,
+    IEEE802154_DSTPAN_OFFSET      = 3,
+    IEEE802154_DSTADDR_OFFSET     = 5,
+
+    IEEE802154_SEC_LEVEL_MASK     = 7 << 0,
+
+    IEEE802154_KEY_ID_MODE_0      = 0 << 3,
+    IEEE802154_KEY_ID_MODE_1      = 1 << 3,
+    IEEE802154_KEY_ID_MODE_2      = 2 << 3,
+    IEEE802154_KEY_ID_MODE_3      = 3 << 3,
+    IEEE802154_KEY_ID_MODE_MASK   = 3 << 3,
+
+    IEEE802154_MACCMD_DATA_REQ    = 4
+};
+
+enum
+{
+    MRF24J40_RSSI_OFFSET = 90,
+    MRF24J40_RSSI_SLOPE = 5
+};
+
+static void radioTransmitMessage(otInstance *aInstance);
+
+static RadioPacket sTransmitFrame;
+static RadioPacket sReceiveFrame;
+static RadioPacket sAckFrame;
+
+static ThreadError sTransmitError;
+static ThreadError sReceiveError;
+
+static uint8_t sTransmitPsdu[IEEE802154_MAX_LENGTH];
+static uint8_t sReceivePsdu[IEEE802154_MAX_LENGTH];
+static uint8_t sAckPsdu[IEEE802154_MAX_LENGTH];
+
+static PhyState sState = kStateDisabled;
+static bool sIsReceiverEnabled = false;
+
+static volatile uint8_t Mrf24StatusTx = 0;
+static volatile uint8_t Mrf24StatusRx = 0;
+static volatile uint8_t Mrf24StatusSec = 0;
+
+static DEV_SPI_PTR pmrf_spi_ptr;
+static DEV_GPIO_PTR pmrf_gpio_ptr;
+static void RadioIsr(void *ptr);
+
+/* Variables for test */
+static uint32_t numInterruptRev = 0;
+static uint32_t numInterruptTrans = 0;
+static uint32_t numRadioProcess = 0;
+static uint32_t NODE_ID = 1;
```

Comment:
Yes. EMSK is a FPGA-based hardware platform. I have rechecked it. There is no unique ID called by software. I am contacting hardware team to send me a way for device ID. I will update it soon.